server/src/main/java/io/crate/auth/CachingJwkProvider.java

```diff
@@ -0,0 +1,220 @@
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.auth;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.MalformedURLException;
+import java.net.Proxy;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.net.URLConnection;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import org.elasticsearch.common.Strings;
+import org.jetbrains.annotations.VisibleForTesting;
+
+import com.auth0.jwk.Jwk;
+import com.auth0.jwk.JwkProvider;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.ObjectReader;
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.Expiry;
+
+import io.netty.handler.codec.http.HttpHeaderNames;
+
+/**
+ * Custom @{@link JwkProvider} implementation which caches results
+ * of public jwk keys for the duration of the "Cache-Control max-age"
+ * Http header valye from the response of the jwt authentication endpoint or
+ * a provided default value.
+ */
+public class CachingJwkProvider implements JwkProvider {
+
+    private final URL url;
+    private final Proxy proxy;
+    private final Map<String, String> headers;
+    private final Integer connectTimeout;
+    private final Integer readTimeout;
+    private final ObjectReader reader;
+    private final Cache<String, JwkResult> cache;
+    private final Duration cacheExpirationTime;
+
+
+    public CachingJwkProvider(String domain, Duration cacheExpirationTime) {
+        this(urlForDomain(domain), cacheExpirationTime, null, null, null, null);
+    }
+
+    public CachingJwkProvider(URL url, Duration cacheExpirationTime, Integer connectTimeout, Integer readTimeout, Proxy proxy, Map<String, String> headers) {
+        if (url == null) {
+            throw new IllegalArgumentException("A non-null url is required");
+        }
+
+        if (connectTimeout != null && connectTimeout < 0) {
+            throw new IllegalArgumentException("Invalid connect timeout value '" + connectTimeout + "'. Must be a non-negative integer.");
+        }
+
+        if (readTimeout != null && readTimeout < 0) {
+            throw new IllegalArgumentException("Invalid read timeout value '" + readTimeout + "'. Must be a non-negative integer.");
+        }
+
+
+        this.url = url;
+        this.proxy = proxy;
+        this.connectTimeout = connectTimeout;
+        this.readTimeout = readTimeout;
+        this.reader = new ObjectMapper().readerFor(Map.class);
+        this.headers = (headers == null) ?
+            Collections.singletonMap("Accept", "application/json") : headers;
+        this.cacheExpirationTime = cacheExpirationTime;
+        this.cache = Caffeine.newBuilder()
```

Comment:
Cool, looks like we can completely strip out guava dependency (maybe during feature freeze). I
t's pulled by the library but with this change will never be used.

In https://github.com/crate/crate/commit/e6f19a20b374920d8f674e57cb9865a1100d9237 we already adding it only in runtime and probably can remove completely - maybe can add TODO for that